src/core/compiler.jl

```diff
@@ -1,150 +1,126 @@
-"""
-function generate_assume(var::Union{Symbol, Expr}, dist, model_info)
-    main_body_names = model_info[:main_body_names]
-    vi = main_body_names[:vi]
-    sampler = main_body_names[:sampler]
-
-    varname = gensym(:varname)
-    sym, idcs, csym = gensym(:sym), gensym(:idcs), gensym(:csym)
-    csym_str, indexing, syms = gensym(:csym_str), gensym(:indexing), gensym(:syms)
-
-    if var isa Symbol
-        varname_expr = quote
-            $sym, $idcs, $csym = Turing.@VarName $var
-            $csym = Symbol($(QuoteNode(model_info[:name])), $csym)
-            $varname = Turing.VarName{$sym}($csym, "")
-        end
-    else
-        varname_expr = quote
-            $sym, $idcs, $csym = Turing.@VarName $var
-            $csym_str = string($(QuoteNode(model_info[:name])))*string($csym)
-            $indexing = foldl(*, $idcs, init = "")
-            $varname = Turing.VarName{$sym}(Symbol($csym_str), $indexing)
-        end
-    end
-
-    lp = gensym(:lp)
-    return quote
-        $varname_expr
-        isdist = if isa($dist, AbstractVector)
-            # Check if the right-hand side is a vector of distributions.
-            all(d -> isa(d, Distribution), $dist)
+function split_var_str(var_str, inds_as = Vector)
+    ind = findfirst(c -> c == '[', var_str)
+    if inds_as === String
+        if ind === nothing
+            return var_str, ""
         else
-            # Check if the right-hand side is a distribution.
-            isa($dist, Distribution)
+            return var_str[1:ind-1], var_str[ind:end]
         end
-        @assert isdist @error($(wrong_dist_errormsg(@__LINE__)))
-
-        ($var, $lp) = if isa($dist, AbstractVector)
-            Turing.assume($sampler, $dist, $varname, $var, $vi)
-        else
-            Turing.assume($sampler, $dist, $varname, $vi)
+    end
+    @assert inds_as === Vector
+    inds = Vector{String}[]
+    if ind === nothing
+        return var_str, inds
+    end
+    sym = var_str[1:ind-1]
+    ind = length(sym)
+    while ind < length(var_str)
+        ind += 1
+        @assert var_str[ind] == '['
+        push!(inds, String[])
+        while var_str[ind] != ']'
+            ind += 1
+            if var_str[ind] == '['
+                ind2 = findnext(c -> c == ']', var_str, ind)
+                push!(inds[end], strip(var_str[ind:ind2]))
+                ind = ind2+1
+            else
+                ind2 = findnext(c -> c == ',' || c == ']', var_str, ind)
+                push!(inds[end], strip(var_str[ind:ind2-1]))
+                ind = ind2
+            end
         end
-        $vi.logp += $lp
     end
+    return sym, inds
 end
 
-"""
-    tilde(left, right, model_info)
-
-The `tilde` function generates observation expression for data variables and assumption expressions for parameter variables, updating `model_info` in the process.
-"""
-function tilde(left, right, model_info)
-    return generate_observe(left, right, model_info)
+function assert_dist(dist; msg)
```

Comment:
Maybe worth refactor this via multiple dispatch.

Reply:
Sure